module_2/gifts/service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java

```diff
@@ -0,0 +1,224 @@
+import com.epam.esm.service.validation.Util;
+
+/**
+ * 
+ * Contains methods implementation for working mostly with {@code CertificateDto}
+ * entity.
+ *
+ */
+@Service
+public class CertificateServiceImpl implements CertificateService {
+
+	@Autowired
+	private CertificateRepository certificateRepository;
+
+	@Autowired
+	private TagRepository tagRepository;
+
+	@Autowired
+	private CertificateValidation certificateValidation;
+
+	@Autowired
+	private TagValidation tagValidation;
+
+	@Autowired
+	private CertificateConverter certificateConverter;
+
+	@Autowired
+	private TagConverter tagConverter;
+
+	@Autowired
+	private TagService tagService;
+
+	/**
+	 * Creates and saves the passed certificate.
+	 * 
+	 * @param certificateDto the certificate to be saved
+	 * @return saved certificate
+	 */
+	@Override
+	@Transactional
+	public CertificateDto create(CertificateDto certificateDto) {
+		certificateValidation.validateCertificateAllFieldsRequirementsForCreate(certificateDto);
+
+		LocalDateTime now = LocalDateTime.now();
+		certificateDto.setCreateDate(now);
+		certificateDto.setLastUpdateDate(now);
+
+		CertificateModel createdCertificateModel = certificateRepository
+				.create(certificateConverter.convertToModel(certificateDto));
+		CertificateDto createdCertificate = certificateConverter.convertToDto(createdCertificateModel);
+		createdCertificate.setTags(saveCertificateTags(createdCertificate.getId(), certificateDto.getTags()));
+
+		return createdCertificate;
+	}
+
+	/**
+	 * Reads certificate with passed id.
+	 * 
+	 * @param certificateId the id of certificate to be read
+	 * @return certificate with passed id
+	 */
+	@Override
+	public CertificateDto readById(long certificateId) {
+		certificateValidation.checkCertificateExistenceById(certificateId);
+
+		CertificateDto certificateDto = certificateConverter
+				.convertToDto(certificateRepository.readById(certificateId));
+
+		certificateDto.setTags(tagService.readByCertificateId(certificateId));
+
+		return certificateDto;
+	}
+
+	/**
+	 * Reads all certificates according to passed parameters.
+	 * 
+	 * @param params the parameters which define choice of certificates and their
+	 *               ordering
+	 * @return certificates which meet passed parameters
+	 */
+	@Override
+	public List<CertificateDto> readAll(MultiValueMap<String, String> params) {
+		MultiValueMap<String, String> paramsInLowerCase = Util.mapToLowerCase(params);
+		certificateValidation.validateReadParams(paramsInLowerCase);
+		List<CertificateModel> certificateModels = certificateRepository.readAll(paramsInLowerCase);
+		List<CertificateDto> certificateDtos = new ArrayList<>(certificateModels.size());
+		for (CertificateModel certificateModel : certificateModels) {
+			CertificateDto certificateDto = certificateConverter.convertToDto(certificateModel);
+			certificateDto.setTags(tagService.readByCertificateId(certificateDto.getId()));
+			certificateDtos.add(certificateDto);
+		}
```

Comment:
Why do you call to sql query instead of use one with join? You make n+1 qieries

Reply:
@maximBortnik, the same situation as for the readById method in the same class: the CertificateModel contains fields only for the certificate database table while CertificateDto contains additional field with the list of tags.